### PR TITLE
Enable basic RadosGW backend

### PR DESCRIPTION
**Pull request by sabaini**
Backend part of the RGW service

Signed-off-by: Peter Sabaini <peter.sabaini@canonical.com>

### CODE REVIEW

File: microceph/ceph/configwriter_test.go [hunk 1]
```diff
@@ -41,9 +41,27 @@ func (s *configWriterSuite) TestWriteCephConfig() {
...
 }
 
+// Test ceph config writing
+func (s *configWriterSuite) TestWriteRadosGWConfig() {
+	config := newRadosGWConfig(s.tmp)
+	err := config.WriteConfig(
+		map[string]any{
+			"monitors": "foohost",
+		},
+	)
```
Comment (stgraber) [inline microceph/ceph/configwriter_test.go:51]: Line break after this

File: microceph/ceph/rgw.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	err := conf.WriteConfig(
+		map[string]any{
+			"runDir":   runPath,
+			"monitors": s.ClusterState().Address().Hostname(),
+			"rgwPort":  port,
+		},
+	)
+	if err != nil {
+		return err
+	}
```
Comment (stgraber) [inline microceph/ceph/rgw.go:30]: Line break after this

File: microceph/ceph/rgw.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	// Create RGW keyring.
+	path := filepath.Join(dataPath, "radosgw", "ceph-radosgw.gateway")
+	if err = createRGWKeyring(path); err != nil {
+		return err
+	}
```
Comment (stgraber) [inline microceph/ceph/rgw.go:35]: Line break after this

File: microceph/ceph/rgw.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	// Symlink the keyring to the conf directory for usage with the radosgw-admin command.
+	if err = symlinkRGWKeyring(path, confPath); err != nil {
+		return err
+	}
```
Comment (stgraber) [inline microceph/ceph/rgw.go:39]: Line break after this

File: microceph/ceph/rgw.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	// Record the changes to the database.
+	if err = rgwUpdateDatabase(s); err != nil {
+		return err
+	}
+	if err = startRGW(); err != nil {
```
Comment (stgraber) [inline microceph/ceph/rgw.go:44]: Line break after this

File: microceph/ceph/rgw.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		return err
+	}
+	return nil
+}
+
+// rgwUpdateDatabase records changes to the database.
+func rgwUpdateDatabase(s common.StateInterface) error {
+	if s.ClusterState().Database == nil {
+		return fmt.Errorf("no database")
+	}
```
Comment (stgraber) [inline microceph/ceph/rgw.go:54]: Line break after this

File: microceph/ceph/rgw.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	err := s.ClusterState().Database.Transaction(s.ClusterState().Context, func(ctx context.Context, tx *sql.Tx) error {
+		// Record the role.
+		_, err := database.CreateService(ctx, tx, database.Service{Member: s.ClusterState().Name(), Service: "rgw"})
+		if err != nil {
+			return fmt.Errorf("Failed to record role: %w", err)
+		}
```
Comment (stgraber) [inline microceph/ceph/rgw.go:60]: Line break after this

File: microceph/ceph/rgw.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+		return nil
+	})
```
Comment (stgraber) [inline microceph/ceph/rgw.go:62]: Line break after this

File: microceph/ceph/rgw.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	return err
+}
+
+// startRGW starts the RGW service.
+func startRGW() error {
+	err := snapStart("rgw", true)
+	if err != nil {
+		return fmt.Errorf("Failed to start RGW service: %w", err)
+	}
```
Comment (stgraber) [inline microceph/ceph/rgw.go:71]: Line break after this

File: microceph/ceph/rgw.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	return nil
+}
+
+// createRGWKeyring creates the RGW keyring.
+func createRGWKeyring(path string) error {
+	if err := os.MkdirAll(path, 0770); err != nil {
+		return err
+	}
```
Comment (stgraber) [inline microceph/ceph/rgw.go:79]: Line break after this

File: microceph/ceph/rgw.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+	return nil
+}
+
+// symlinkRGWKeyring creates a symlink to the RGW keyring in the conf directory for use with the radosgw-admin command.
+func symlinkRGWKeyring(keyPath, confPath string) error {
+	if err := os.Symlink(
+		filepath.Join(keyPath, "keyring"),
+		filepath.Join(confPath, "ceph.client.radosgw.gateway.keyring")); err != nil {
+		return fmt.Errorf("Failed to create symlink to RGW keyring: %w", err)
+	}
```
Comment (stgraber) [inline microceph/ceph/rgw.go:103]: Line break after this

**Review by stgraber — COMMENTED** [review]
Just the same code style nit.

As this project was moved over to the Ceph team, please let me know if this isn't something you care about and I'll stop flagging those in reviews ;)

The code style in here would not be acceptable in a LXD/MicroCloud PR, but it's not invalid Go syntax and certainly is allowed or even recommended by other Go projects.

**Comment by sabaini** [comment]
I am fine with either code style really, but it'd be great to have something that either formats this automatically or lints for it. I don't think gofmt does? Do you know of some way to automate this?

Thanks!

**Comment by stgraber** [comment]
https://github.com/lxc/lxd/blob/master/test/lint/newline-after-block.sh

**Comment by sabaini** [comment]
Thanks for the linter scripts, those look useful. The newline-after-block.sh linter moans about quite a lot of violations in the current codebase indeed but I guess these should go a in a separate PR. 

I have pushed an update for style (note though the linter did not mark [configwriter_test.go](https://github.com/canonical/microceph/pull/84/files#diff-59de60dd0c5621487b275e4d4fc5a3434b2855e9f9022606feecbc7dbbd716e6), afaics because it matches for a lone `}` but not `},`) -- from my perspective I'd like to use the linters as the (executable) style spec as much as possible, i.e. things that the linter does not catch are per definition ok style.

**Pull request merged**